### PR TITLE
Add c-sharp textobjects

### PR DESCRIPTION
**Pull request by votagus**
This PR adds basic textobjext based navigation for c-sharp.

### CODE REVIEW

**Review by the-mikedavis — COMMENTED** [review]
Jumping around in some C# files I have locally, I think we should also have function textobjects for property definitions like:

```csharp
public override bool CanRead {
	get { return true; }
}
```

But I'm not sure if we should tag the entire property as `function.around` (the `property_declaration` node) or just the accessor declaration (`accessor_declaration` node)

Also constructor definitions:

```csharp
public BufferPoolStream(BufferPool bufferPool) {
	if (bufferPool == null)
		throw new ArgumentNullException("bufferPool");
	_bufferPool = bufferPool;
}
```

```scm
(constructor_declaration body: (_) @function.inside) @function.around
```

**Comment by votagus** [comment]
Do you think it would be appropriate to map it like this until we have more precise navigation commands?
```
([
  (class_declaration body: (_) @class.inside)
  (struct_declaration body: (_) @class.inside)
  (interface_declaration body: (_) @class.inside)
  (enum_declaration body: (_) @class.inside)
  (delegate_declaration)
  (record_declaration body: (_) @class.inside)
  (record_struct_declaration body: (_) @class.inside)
  ]) @class.around
```
**Note: To make `record_struct_declaration` work the grammar revision has to be updated along with its highlights.scm.**

About `property_declaration` vs `accessor_declaration`, playing around, `(property_declaration (_) @function.inside) @function.around` _feels_ more consistent with the rest of the commands, specially when using it with match inside/around function (`mif`, `maf`).

Class for testing.
```csharp
namespace NavigationTest {
	public class ClassTest {
		public int field = 0;
		
		public int Field {
			get { return field; }
			set { field = value; }
		}

		public NavigationTest(string param1, int param2) {
			// Set field.
			field = param2;
		}

		~NavigationTest() {
			// Set field.
			field = param2;
		}

		/// Documentation.
		///
		/// Documentation.
		public int Method(string param1, int param2) {
			/* Multiline
				comment.
			*/
			return 0;
		}
	}
	
	public enum EnumTest {
		Item1,
		Item2
	}
	
	public interface InterfaceTest {
		public void Method();
	}
	
	public struct StructTest {
		public int field = 0;
	}

	public record RecordTest {
		public int field = 0;
	}

	public record struct RecordStructTest {
		public int field = 0;
	}
}
```

Let me know what you think.

File: runtime/queries/c-sharp/textobjects.scm [hunk 1]
```diff
@@ -0,0 +1,21 @@
+([
+  (class_declaration body: (_) @class.inside)
+  (struct_declaration body: (_) @class.inside)
+  (interface_declaration body: (_) @class.inside)
+  (enum_declaration body: (_) @class.inside)
+  (delegate_declaration)
+  (record_declaration body: (_) @class.inside)
+  (record_struct_declaration body: (_) @class.inside)
+  ]) @class.around
```
Comment (the-mikedavis) [inline runtime/queries/c-sharp/textobjects.scm:9]:
```suggestion
[
  (class_declaration body: (_) @class.inside)
  (struct_declaration body: (_) @class.inside)
  (interface_declaration body: (_) @class.inside)
  (enum_declaration body: (_) @class.inside)
  (delegate_declaration)
  (record_declaration body: (_) @class.inside)
  (record_struct_declaration body: (_) @class.inside)
] @class.around
```

File: runtime/queries/c-sharp/highlights.scm [hunk 2]
```diff
@@ -1,47 +1,17 @@
 ;; Methods
-(method_declaration
```
Comment (the-mikedavis) [inline runtime/queries/c-sharp/highlights.scm:2]: Most of the changes that we have in highlights.scm are improvements that haven't been upstreamed. Can you revert the changes to that file? It looks like the only changes necessary for updating the grammar come from the preprocessor change in https://github.com/tree-sitter/tree-sitter-c-sharp/pull/191:

```diff
diff --git a/runtime/queries/c-sharp/highlights.scm b/runtime/queries/c-sharp/highlights.scm
index 6b94dd8c..05eda6b9 100644
--- a/runtime/queries/c-sharp/highlights.scm
+++ b/runtime/queries/c-sharp/highlights.scm
@@ -228,8 +228,21 @@
   "let"
 ] @keyword
 
-(preprocessor_call) @keyword.directive
-(preprocessor_call (_) @keyword.directive)
+[
+  (if_directive (_) @keyword.directive)
+  (elif_directive (_) @keyword.directive)
+  (else_directive)
+  (endif_directive)
+  (warning_directive (_) @keyword.directive)
+  (error_directive (_) @keyword.directive)
+  (line_directive (_)? @keyword.directive)
+  (nullable_directive)
+  (define_directive (_) @keyword.directive)
+  (undef_directive (_) @keyword.directive)
+  (region_directive)
+  (endregion_directive)
+  (pragma_directive)
+] @keyword.directive
 
 ;; Linq
 (from_clause (identifier) @variable)
```

Comment (votagus) [reply runtime/queries/c-sharp/highlights.scm:2]: Yeah, I just found out that by looking at the commits, I wrongly assumed that this file was in sync with the main repo, I'll try to add those directives as well.

**Review by the-mikedavis — APPROVED** [review]

**Comment by the-mikedavis** [comment]
Looks great, thank you!
